abe/abe.go

```diff
@@ -0,0 +1,472 @@
+			found = true
+			break
+		}
+		if numBrc == 0 && i < len(boolExp)-2 && boolExp[i:i+2] == "OR" {
+			boolExp1 = boolExp[:i]
+			boolExp2 = boolExp[i+2:]
+			msp1, c1, err = booleanToMspIterative(boolExp1, vec, c)
+			if err != nil {
+				return nil, 0, err
+			}
+			msp2, cOut, err = booleanToMspIterative(boolExp2, vec, c1)
+			if err != nil {
+				return nil, 0, err
+			}
+			found = true
+			break
+		}
+	}
+
+	// If the AND or OR gate is not found then there are two options,
+	// ether the whole expression is in brackets, or the the expression
+	// is only one attribute. It nether of both is true, then
+	// an error is returned while converting the expression into an
+	// attribute
+	if found == false {
+		if boolExp[0] == '(' && boolExp[len(boolExp)-1] == ')' {
+			boolExp = boolExp[1:(len(boolExp) - 1)]
+			return booleanToMspIterative(boolExp, vec, c)
+		}
+
+		attrib, err := strconv.Atoi(boolExp)
+		if err != nil {
+			return nil, 0, err
+		}
+		mat := make(data.Matrix, 1)
+		mat[0] = make(data.Vector, c)
+		for i := 0; i < c; i++ {
+			if i < len(vec) {
+				mat[0][i] = new(big.Int).Set(vec[i])
+			} else {
+				mat[0][i] = big.NewInt(0)
+			}
+		}
+
+		rowToAttrib := make([]int, 1)
+		rowToAttrib[0] = attrib
+		return &Msp{mat: mat, rowToAttrib: rowToAttrib}, c, nil
+	} else {
+		// otherwise we join the two msp structures into one
+		mat := make(data.Matrix, len(msp1.mat)+len(msp2.mat))
+		for i := 0; i < len(msp1.mat); i++ {
+			mat[i] = make(data.Vector, cOut)
+			for j := 0; j < len(msp1.mat[0]); j++ {
+				mat[i][j] = msp1.mat[i][j]
+			}
+			for j := len(msp1.mat[0]); j < cOut; j++ {
+				mat[i][j] = big.NewInt(0)
+			}
+		}
+		for i := 0; i < len(msp2.mat); i++ {
+			mat[i+len(msp1.mat)] = msp2.mat[i]
+		}
+		rowToAttrib := append(msp1.rowToAttrib, msp2.rowToAttrib...)
+
+		return &Msp{mat: mat, rowToAttrib: rowToAttrib}, cOut, nil
+	}
+}
+
+// makeAndVecs is a helping structure that given a vector and and counter
+// creates two new vectors used whenever an AND gate is found in a iterative
+// step of BooleanToMsp
+func makeAndVecs(vec data.Vector, c int) (data.Vector, data.Vector) {
+	vec1 := make(data.Vector, c+1)
+	vec2 := make(data.Vector, c+1)
+	for i := 0; i < c; i++ {
+		vec1[i] = big.NewInt(0)
+		if i < len(vec) {
+			vec2[i] = new(big.Int).Set(vec[i])
+		} else {
+			vec2[i] = big.NewInt(0)
+		}
+	}
+	vec1[c] = big.NewInt(-1)
+	vec2[c] = big.NewInt(1)
+
+	return vec1, vec2
+}
+
+// gaussianElimination solves a vector equation mat * x = v and finds vector x,
+// using Gaussian elimination. Arithmetic operations are considered to be over
+// Z_p, where p should be a prime number. If such x does not exist, then the
+// function returns an error.
+func gaussianElimination(mat data.Matrix, v data.Vector, p *big.Int) (data.Vector, error) {
+	if len(mat) == 0 || len(mat[0]) == 0 {
+		return nil, fmt.Errorf("the matrix should not be empty")
+	}
+	if len(mat) != len(v) {
+		return nil, fmt.Errorf("dimensions should match")
```

Comment:
Could you change the error message to also state the actual dimensions that mismatch?